File: server/feed12/app/info/glennengstrand/news/dao/InboundDao.scala
package info.glennengstrand.news.dao

import javax.inject.{Inject, Singleton}

import akka.actor.ActorSystem
import play.api.libs.concurrent.CustomExecutionContext
import play.api.{Logger, MarkerContext}
import scala.jdk.CollectionConverters._

import com.datastax.oss.driver.api.core.cql.{BoundStatement, PreparedStatement}
import com.datastax.oss.driver.api.core.CqlSession

import scala.concurrent.Future
import info.glennengstrand.news.model._

class InboundExecutionContext @Inject()(actorSystem: ActorSystem)
    extends CustomExecutionContext(actorSystem, "repository.dispatcher")

trait InboundDao {
  def create(id: Int, data: Inbound)(implicit mc: MarkerContext): Future[Inbound]

  def get(id: Int)(implicit mc: MarkerContext): Future[Seq[Inbound]]
}

@Singleton
class InboundDaoImpl @Inject()(nosql: NoSqlDao)(implicit ec: InboundExecutionContext)
    extends InboundDao with Link {

  private val logger = Logger(this.getClass)
  private val selectCql = "select toTimestamp(occurred) as occurred, fromparticipantid, subject, story from Inbound where participantid = ? order by occurred desc"
  private val insertCql = "insert into Inbound (ParticipantID, FromParticipantID, Occurred, Subject, Story) values (?, ?, now(), ?, ?)"
  private lazy val session: CqlSession = connect
  private lazy val insertStmt = session.prepare(insertCql)
  private lazy val selectStmt = session.prepare(selectCql)

  private def connect: CqlSession = {
    var rv: CqlSession = null
    nosql.connect(s => rv = s)
    rv
  }
  
  override def get(id: Int)(
      implicit mc: MarkerContext): Future[Seq[Inbound]] = {
    Future {
      val bs = selectStmt.bind(id.asInstanceOf[Object])
      val retVal = for {
        r <- session.execute(bs).iterator().asScala
      } yield Inbound(Option(toLink(r.getInt(1).toLong)), Option(toLink(id.toLong)), Option(r.getInstant(0).toString()), Option(r.getString(2)), Option(r.getString(3)))
      retVal.toSeq
    }
  }

  def create(id: Int, data: Inbound)(implicit mc: MarkerContext): Future[Inbound] = {
    Future {
      val bs = insertStmt.bind(new java.lang.Integer(extractId(data.to.get.toString.asInstanceOf[String]).toInt), 
          new java.lang.Integer(extractId(data.from.get.toString.asInstanceOf[String]).toInt), 
          data.subject.get.toString.asInstanceOf[String], 
          data.story.get.toString.asInstanceOf[String])
      session.execute(bs)
      data
    }
  }

}

object MockInboundDaoImpl {
  val test = Inbound(
      from = Option("/participant/1"),
      to = Option("/participant/2"),
      occurred = Option("2020/06/14"),
      subject = Option("test subject"),
      story = Option("test story")
  )
}

@Singleton
class MockInboundDaoImpl @Inject()()(implicit ec: InboundExecutionContext)
    extends InboundDao {

  override def get(id: Int)(
      implicit mc: MarkerContext): Future[Seq[Inbound]] = {
    Future {
      Seq(MockInboundDaoImpl.test)
    }
  }

  override def create(id: Int, data: Inbound)(implicit mc: MarkerContext): Future[Inbound] = {
    Future {
      data
    }
  }

}

File: server/feed15/src/main/kotlin/info/glennengstrand/newsfeed/daos/InboundDao.kt
package info.glennengstrand.newsfeed.daos

import com.datastax.oss.driver.api.core.CqlSession
import com.datastax.oss.driver.api.core.CqlSessionBuilder
import com.datastax.oss.driver.api.core.cql.PreparedStatement
import info.glennengstrand.newsfeed.models.InboundModel
import info.glennengstrand.newsfeed.models.ParticipantModel
import mu.KotlinLogging
import org.springframework.stereotype.Component
import reactor.core.publisher.Mono
import java.util.concurrent.CompletableFuture

@Component
class InboundDao {
    private val n = NoSqlDao()
    private val logger = KotlinLogging.logger {}
    private val selectCql = """select toTimestamp(occurred) as occurred, fromparticipantid, subject, story 
    from Inbound where participantid = ? order by occurred desc"""
    private val insertCql = """insert into Inbound (ParticipantID, FromParticipantID, Occurred, Subject, Story) 
    values (?, ?, now(), ?, ?)"""
    private val selectStatement: PreparedStatement by lazy {
        session.prepare(selectCql)
    }
    private val insertStatement: PreparedStatement by lazy {
        session.prepare(insertCql)
    }
    private val session: CqlSession by lazy {
        var rv: CqlSessionBuilder? = null
        n.connect {
            rv = it
        }
        rv?.build()!!
    }

    fun getInbound(id: Long): Mono<List<InboundModel>> {
        val from = ParticipantModel(id, "").link
        return Mono.fromFuture {
            CompletableFuture.supplyAsync<List<InboundModel>> {
                val bs = selectStatement.bind(id.toInt())
                val rs = session.execute(bs)
                val rv = mutableListOf<InboundModel>()
                rs.forEach {
                    rv.add(
                        InboundModel(
                            ParticipantModel(it.getInt(1).toLong(), "").link,
                            from,
                            n.format(it.getInstant(0)),
                            it.getString(2)!!,
                            it.getString(3)!!,
                        ),
                    )
                }
                rv
            }
        }
    }

    fun addInbound(
        id: Long,
        ib: InboundModel,
    ): Mono<InboundModel> {
        val bs =
            insertStatement.bind(
                ParticipantModel.unlink(ib.to).toInt(),
                id.toInt(),
                ib.subject,
                ib.story,
            )
        session.executeAsync(bs)
        return Mono.just(ib)
    }
}
